File: Backend/src/services/cloudinaryService.js
/**
 * Cloudinary Service - Centralized Image Management
 * 
 * This service handles all image operations:
 * - Upload (from file, base64, URL)
 * - Delete
 * - Transform (resize, crop, format)
 * - Get optimized URLs
 * 
 * Usage Examples:
 * - Profile pictures
 * - National ID documents
 * - Medical records/PDFs
 * - Any other images in the app
 */

// Note: dotenv is already loaded in server.js
const cloudinary = require('cloudinary').v2;

// Configure Cloudinary
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
  secure: true,
});

// Log Cloudinary configuration status
console.log('☁️ Cloudinary Config:', {
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME ? '✅ ' + process.env.CLOUDINARY_CLOUD_NAME : '❌ Missing',
  api_key: process.env.CLOUDINARY_API_KEY ? '✅ Set' : '❌ Missing',
  api_secret: process.env.CLOUDINARY_API_SECRET ? '✅ Set' : '❌ Missing',
});

/**
 * Folder structure in Cloudinary:
 * housepital/
 *   ├── profiles/          - User profile pictures
 *   ├── id-documents/      - National ID cards (front/back)
 *   ├── medical-records/   - Medical documents, PDFs
 *   ├── prescriptions/     - Prescription images
 *   └── general/           - Other images
 */
const FOLDERS = {
  PROFILES: 'housepital/profiles',
  ID_DOCUMENTS: 'housepital/id-documents',
  MEDICAL_RECORDS: 'housepital/medical-records',
  PRESCRIPTIONS: 'housepital/prescriptions',
  GENERAL: 'housepital/general',
};

/**
 * Upload image from base64 string
 * @param {string} base64String - The base64 encoded image (with or without data URI prefix)
 * @param {object} options - Upload options
 * @param {string} options.folder - Cloudinary folder (use FOLDERS constant)
 * @param {string} options.publicId - Custom public ID (optional)
 * @param {string} options.resourceType - 'image', 'raw', or 'auto' (default: 'auto')
 * @param {object} options.transformation - Transformation options
 * @returns {Promise<object>} - Upload result with url and publicId
 */
const uploadFromBase64 = async (base64String, options = {}) => {
  try {
    const {
      folder = FOLDERS.GENERAL,
      publicId = null,
      resourceType = 'auto',
      transformation = null,
    } = options;

    // Add data URI prefix if not present
    let dataUri = base64String;
    if (!base64String.startsWith('data:')) {
      // Try to detect the image type from base64
      const signatures = {
        '/9j/': 'image/jpeg',
        'iVBORw0KGgo': 'image/png',
        'R0lGODlh': 'image/gif',
        'UklGR': 'image/webp',
        'JVBERi0': 'application/pdf',
      };
      
      let mimeType = 'image/jpeg'; // default
      for (const [signature, mime] of Object.entries(signatures)) {
        if (base64String.startsWith(signature)) {
          mimeType = mime;
          break;
        }
      }
      dataUri = `data:${mimeType};base64,${base64String}`;
    }

    const uploadOptions = {
      folder,
      resource_type: resourceType,
      ...(publicId && { public_id: publicId }),
      ...(transformation && { transformation }),
    };

    console.log('📤 Uploading to Cloudinary...', { folder, resourceType });
    const result = await cloudinary.uploader.upload(dataUri, uploadOptions);
    console.log('✅ Upload successful:', { url: result.secure_url, publicId: result.public_id });

    return {
      success: true,
      url: result.secure_url,
      publicId: result.public_id,
      format: result.format,
      width: result.width,
      height: result.height,
      bytes: result.bytes,
      resourceType: result.resource_type,
    };
  } catch (error) {
    console.error('❌ Cloudinary upload error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Upload image from buffer (for multer uploads)
 * @param {Buffer} buffer - The file buffer
 * @param {object} options - Upload options
 * @returns {Promise<object>} - Upload result
 */
const uploadFromBuffer = async (buffer, options = {}) => {
  try {
    console.log('📤 uploadFromBuffer called, buffer size:', buffer?.length);
    
    const {
      folder = FOLDERS.GENERAL,
      publicId = null,
      resourceType = 'auto',
      transformation = null,
    } = options;

    console.log('📂 Upload options:', { folder, publicId, resourceType });

    const result = await new Promise((resolve, reject) => {
      const uploadOptions = {
        folder,
        resource_type: resourceType,
        ...(publicId && { public_id: publicId }),
        ...(transformation && { transformation }),
      };

      console.log('☁️ Calling cloudinary.uploader.upload_stream...');
      const stream = cloudinary.uploader.upload_stream(
        uploadOptions,
        (error, result) => {
          if (error) {
            console.log('❌ Stream error:', error);
            reject(error);
          } else {
            console.log('✅ Stream success:', result?.secure_url);
            resolve(result);
          }
        }
      );
      stream.end(buffer);
    });

    console.log('✅ Upload complete:', result.secure_url);

    return {
      success: true,
      url: result.secure_url,
      publicId: result.public_id,
      format: result.format,
      width: result.width,
      height: result.height,
      bytes: result.bytes,
      resourceType: result.resource_type,
    };
  } catch (error) {
    console.error('❌ Cloudinary buffer upload error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Upload image from URL
 * @param {string} imageUrl - The URL of the image to upload
 * @param {object} options - Upload options
 * @returns {Promise<object>} - Upload result
 */
const uploadFromUrl = async (imageUrl, options = {}) => {
  try {
    const {
      folder = FOLDERS.GENERAL,
      publicId = null,
      resourceType = 'auto',
    } = options;

    const result = await cloudinary.uploader.upload(imageUrl, {
      folder,
      resource_type: resourceType,
      ...(publicId && { public_id: publicId }),
    });

    return {
      success: true,
      url: result.secure_url,
      publicId: result.public_id,
      format: result.format,
    };
  } catch (error) {
    console.error('Cloudinary URL upload error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Delete image from Cloudinary
 * @param {string} publicId - The public ID of the image to delete
 * @param {string} resourceType - 'image', 'raw', or 'video' (default: 'image')
 * @returns {Promise<object>} - Deletion result
 */
const deleteImage = async (publicId, resourceType = 'image') => {
  try {
    const result = await cloudinary.uploader.destroy(publicId, {
      resource_type: resourceType,
    });

    return {
      success: result.result === 'ok',
      result: result.result,
    };
  } catch (error) {
    console.error('Cloudinary delete error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Delete multiple images from Cloudinary
 * @param {string[]} publicIds - Array of public IDs to delete
 * @param {string} resourceType - 'image', 'raw', or 'video' (default: 'image')
 * @returns {Promise<object>} - Deletion result
 */
const deleteMultipleImages = async (publicIds, resourceType = 'image') => {
  try {
    const result = await cloudinary.api.delete_resources(publicIds, {
      resource_type: resourceType,
    });

    return {
      success: true,
      deleted: result.deleted,
      notFound: result.not_found || [],
    };
  } catch (error) {
    console.error('Cloudinary bulk delete error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Get optimized/transformed image URL
 * @param {string} publicId - The public ID of the image
 * @param {object} transformations - Transformation options
 * @returns {string} - Transformed image URL
 */
const getTransformedUrl = (publicId, transformations = {}) => {
  const {
    width = null,
    height = null,
    crop = 'fill',
    quality = 'auto',
    format = 'auto',
    effect = null,
  } = transformations;

  const options = {
    quality,
    fetch_format: format,
    ...(width && { width }),
    ...(height && { height }),
    ...(width || height ? { crop } : {}),
    ...(effect && { effect }),
  };

  return cloudinary.url(publicId, options);
};

/**
 * Get thumbnail URL
 * @param {string} publicId - The public ID of the image
 * @param {number} size - Thumbnail size (default: 150)
 * @returns {string} - Thumbnail URL
 */
const getThumbnailUrl = (publicId, size = 150) => {
  return cloudinary.url(publicId, {
    width: size,
    height: size,
    crop: 'thumb',
    gravity: 'face',
    quality: 'auto',
    fetch_format: 'auto',
  });
};

/**
 * Get profile picture URL with face detection
 * @param {string} publicId - The public ID of the image
 * @param {number} size - Image size (default: 200)
 * @returns {string} - Optimized profile picture URL
 */
const getProfilePictureUrl = (publicId, size = 200) => {
  return cloudinary.url(publicId, {
    width: size,
    height: size,
    crop: 'fill',
    gravity: 'face',
    quality: 'auto',
    fetch_format: 'auto',
    radius: 'max', // Circular
  });
};

/**
 * Get ID document URL (optimized for readability)
 * @param {string} publicId - The public ID of the image
 * @returns {string} - Optimized ID document URL
 */
const getIdDocumentUrl = (publicId) => {
  return cloudinary.url(publicId, {
    quality: 'auto:best',
    fetch_format: 'auto',
  });
};

/**
 * Replace an existing image (update)
 * @param {string} publicId - The public ID of the image to replace
 * @param {string} newBase64 - The new image in base64
 * @returns {Promise<object>} - Upload result
 */
const replaceImage = async (publicId, newBase64) => {
  try {
    // Delete old image
    await deleteImage(publicId);
    
    // Upload new image with same public ID structure
    const folder = publicId.substring(0, publicId.lastIndexOf('/'));
    const result = await uploadFromBase64(newBase64, { folder });
    
    return result;
  } catch (error) {
    console.error('Cloudinary replace error:', error);
    return {
      success: false,
      error: error.message,
    };
  }
};

/**
 * Check if Cloudinary is properly configured
 * @returns {Promise<boolean>} - Configuration status
 */
const checkConfiguration = async () => {
  try {
    await cloudinary.api.ping();
    return true;
  } catch (error) {
    console.error('Cloudinary configuration error:', error);
    return false;
  }
};

/**
 * Extract public ID from Cloudinary URL
 * @param {string} url - Cloudinary URL
 * @returns {string|null} - Public ID or null
 */
const extractPublicId = (url) => {
  try {
    if (!url || !url.includes('cloudinary.com')) return null;
    
    // Extract the public ID from URL
    // Format: https://res.cloudinary.com/{cloud_name}/image/upload/{version}/{public_id}.{format}
    const regex = /\/upload\/(?:v\d+\/)?(.+)\.[a-zA-Z]+$/;
    const match = url.match(regex);
    
    return match ? match[1] : null;
  } catch (error) {
    return null;
  }
};

module.exports = {
  // Constants
  FOLDERS,
  
  // Upload methods
  uploadFromBase64,
  uploadFromBuffer,
  uploadFromUrl,
  
  // Delete methods
  deleteImage,
  deleteMultipleImages,
  
  // URL methods
  getTransformedUrl,
  getThumbnailUrl,
  getProfilePictureUrl,
  getIdDocumentUrl,
  
  // Update methods
  replaceImage,
  
  // Utility methods
  checkConfiguration,
  extractPublicId,
  
  // Direct cloudinary access (for advanced usage)
  cloudinary,
};

File: housepital/lib/features/auth/data/datasources/cloudinary_service.dart
// Cloudinary Service - Flutter Client
//
// This service handles all image operations with the backend:
// - Upload images (camera, gallery, base64)
// - Delete images
// - Get transformed URLs
//
// Usage:
// ```dart
// final cloudinaryService = CloudinaryService();
// final result = await cloudinaryService.uploadBase64(base64Image, folder: CloudinaryFolder.idDocuments);
// ```

import 'dart:convert';
import 'dart:io';
import '../../../../core/network/api_service.dart';
import '../../../../core/network/api_constants.dart';

/// Folder types for organizing uploads
enum CloudinaryFolder {
  profiles,
  idDocuments,
  medicalRecords,
  prescriptions,
  general,
}

/// Extension to convert folder enum to string
extension CloudinaryFolderExtension on CloudinaryFolder {
  String get value {
    switch (this) {
      case CloudinaryFolder.profiles:
        return 'profiles';
      case CloudinaryFolder.idDocuments:
        return 'id_documents';
      case CloudinaryFolder.medicalRecords:
        return 'medical_records';
      case CloudinaryFolder.prescriptions:
        return 'prescriptions';
      case CloudinaryFolder.general:
        return 'general';
    }
  }
}

/// Response model for upload operations
class CloudinaryUploadResponse {
  final bool success;
  final String? url;
  final String? publicId;
  final String? format;
  final int? width;
  final int? height;
  final int? size;
  final String? error;

  CloudinaryUploadResponse({
    required this.success,
    this.url,
    this.publicId,
    this.format,
    this.width,
    this.height,
    this.size,
    this.error,
  });

  factory CloudinaryUploadResponse.fromJson(Map<String, dynamic> json) {
    final data = json['data'] as Map<String, dynamic>?;
    return CloudinaryUploadResponse(
      success: json['success'] ?? false,
      url: data?['url'],
      publicId: data?['publicId'],
      format: data?['format'],
      width: data?['width'],
      height: data?['height'],
      size: data?['size'],
      error: json['message'],
    );
  }

  factory CloudinaryUploadResponse.error(String message) {
    return CloudinaryUploadResponse(success: false, error: message);
  }
}

/// Response model for delete operations
class CloudinaryDeleteResponse {
  final bool success;
  final String? message;

  CloudinaryDeleteResponse({required this.success, this.message});

  factory CloudinaryDeleteResponse.fromJson(Map<String, dynamic> json) {
    return CloudinaryDeleteResponse(
      success: json['success'] ?? false,
      message: json['message'],
    );
  }
}

/// Cloudinary Service for handling image uploads
class CloudinaryService {
  final ApiService _apiService;

  CloudinaryService({ApiService? apiService})
    : _apiService = apiService ?? ApiService();

  /// Upload base64 encoded image
  ///
  /// [base64Image] - The base64 string (with or without data URI prefix)
  /// [folder] - Target folder in Cloudinary
  /// [publicId] - Optional custom public ID
  Future<CloudinaryUploadResponse> uploadBase64(
    String base64Image, {
    CloudinaryFolder folder = CloudinaryFolder.general,
    String? publicId,
  }) async {
    try {
      final body = <String, dynamic>{
        'image': base64Image,
        'folder': folder.value,
      };
      if (publicId != null) body['publicId'] = publicId;

      final response = await _apiService.post(
        ApiConstants.cloudinaryUploadBase64,
        body: body,
      );

      return CloudinaryUploadResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryUploadResponse.error(e.toString());
    }
  }

  /// Upload image file (from camera or gallery) - Multipart upload (Recommended)
  ///
  /// [file] - The image file
  /// [folder] - Target folder in Cloudinary
  Future<CloudinaryUploadResponse> uploadFile(
    File file, {
    CloudinaryFolder folder = CloudinaryFolder.general,
  }) async {
    try {
      final response = await _apiService.uploadFile(
        ApiConstants.cloudinaryUpload,
        file,
        fieldName: 'file',
        extraFields: {'folder': folder.value},
      );

      return CloudinaryUploadResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryUploadResponse.error(e.toString());
    }
  }

  /// Upload image from URL
  ///
  /// [imageUrl] - The source image URL
  /// [folder] - Target folder in Cloudinary
  Future<CloudinaryUploadResponse> uploadFromUrl(
    String imageUrl, {
    CloudinaryFolder folder = CloudinaryFolder.general,
  }) async {
    try {
      final response = await _apiService.post(
        '${ApiConstants.cloudinaryBase}/upload-url',
        body: {'url': imageUrl, 'folder': folder.value},
      );

      return CloudinaryUploadResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryUploadResponse.error(e.toString());
    }
  }

  /// Delete single image using POST (since our ApiService.delete doesn't support body)
  ///
  /// [publicId] - The Cloudinary public ID of the image
  Future<CloudinaryDeleteResponse> deleteImage(String publicId) async {
    try {
      // Using POST with action=delete since standard DELETE doesn't support body
      final response = await _apiService.post(
        ApiConstants.cloudinaryDelete,
        body: {'publicId': publicId},
      );

      return CloudinaryDeleteResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryDeleteResponse(success: false, message: e.toString());
    }
  }

  /// Delete multiple images
  ///
  /// [publicIds] - List of Cloudinary public IDs
  Future<CloudinaryDeleteResponse> deleteMultipleImages(
    List<String> publicIds,
  ) async {
    try {
      final response = await _apiService.post(
        '${ApiConstants.cloudinaryBase}/delete-multiple',
        body: {'publicIds': publicIds},
      );

      return CloudinaryDeleteResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryDeleteResponse(success: false, message: e.toString());
    }
  }

  /// Replace existing image with new one
  ///
  /// [publicId] - The public ID of image to replace
  /// [newBase64Image] - The new image in base64
  Future<CloudinaryUploadResponse> replaceImage(
    String publicId,
    String newBase64Image,
  ) async {
    try {
      final response = await _apiService.put(
        '${ApiConstants.cloudinaryBase}/replace',
        body: {'publicId': publicId, 'image': newBase64Image},
      );

      return CloudinaryUploadResponse.fromJson(
        response as Map<String, dynamic>,
      );
    } catch (e) {
      return CloudinaryUploadResponse.error(e.toString());
    }
  }

  /// Get transformed URL for an image
  ///
  /// [publicId] - The Cloudinary public ID
  /// [width] - Target width
  /// [height] - Target height
  Future<String?> getTransformedUrl(
    String publicId, {
    int? width,
    int? height,
    String crop = 'fill',
    String quality = 'auto',
  }) async {
    try {
      String url =
          '${ApiConstants.cloudinaryBase}/transform?publicId=$publicId';
      if (width != null) url += '&width=$width';
      if (height != null) url += '&height=$height';
      url += '&crop=$crop&quality=$quality';

      final response = await _apiService.get(url);
      final data = response as Map<String, dynamic>;

      if (data['success'] == true) {
        return data['url'] as String?;
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  /// Get thumbnail URL
  ///
  /// [publicId] - The Cloudinary public ID
  /// [size] - Thumbnail size (default: 150)
  Future<String?> getThumbnailUrl(String publicId, {int size = 150}) async {
    try {
      final response = await _apiService.get(
        '${ApiConstants.cloudinaryBase}/thumbnail?publicId=$publicId&size=$size',
      );
      final data = response as Map<String, dynamic>;

      if (data['success'] == true) {
        return data['url'] as String?;
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  /// Check Cloudinary service status
  Future<bool> checkStatus() async {
    try {
      final response = await _apiService.get(
        '${ApiConstants.cloudinaryBase}/status',
      );
      final data = response as Map<String, dynamic>;

      return data['configured'] == true;
    } catch (e) {
      return false;
    }
  }

  /// Helper: Convert File to base64
  static Future<String> fileToBase64(File file) async {
    final bytes = await file.readAsBytes();
    return base64Encode(bytes);
  }

  /// Helper: Get file extension from path
  static String getFileExtension(String path) {
    return path.split('.').last.toLowerCase();
  }

  /// Helper: Check if file is valid image
  static bool isValidImage(String path) {
    final validExtensions = [
      'jpg',
      'jpeg',
      'png',
      'gif',
      'webp',
      'heic',
      'heif',
    ];
    final extension = getFileExtension(path);
    return validExtensions.contains(extension);
  }

  /// Helper: Check if file is PDF
  static bool isPdf(String path) {
    return getFileExtension(path) == 'pdf';
  }
}
